packages/compose/src/utils/create-higher-order-component/index.ts

```diff
@@ -0,0 +1,48 @@
+/**
+ * External dependencies
+ */
+import { camelCase, upperFirst } from 'lodash';
+// eslint-disable-next-line no-restricted-imports
+import type { ComponentType } from 'react';
+
+/**
+ * Higher order components can cause props to be obviated. For example a HOC that
+ * injects i18n props will obviate the need for the i18n props to be passed to the component.
+ *
+ * If a HOC does not obviate the need for any specific props then we default to `{}` which
+ * essentially subtracts 0 from the original props of the passed in component. An example
+ * of this is the `pure` HOC which does not change the API surface of the component but
+ * simply modifies the internals.
+ */
+export type HigherOrderComponent< TInnerProps, TOuterProps > = (
+	Inner: ComponentType< TInnerProps >
+) => ComponentType< TOuterProps >;
+
+export type SimpleHigherOrderComponent = < TProps >(
+	Inner: ComponentType< TProps >
+) => ComponentType< TProps >;
```

Comment:
I wish this type could have a better name (what exactly is simple about it?) or if we could avoid it.

If `HigherOrderComponent` defaulted its second argument:
```ts
HigherOrderComponent< TInnerProps, TOuterProps = TInnerProps >
```
then we could write just
```ts
HigherOrderComponent< TInnerProps >
```
to declare a "simple" HOC.

Reply:
I'll do this in the next PR. I agree I hate the names I chose!

Reply:
OK and thanks for working on this. React HOCs are a really complex TypeScript puzzle. The types for React Redux `connect` are almost 700 lines of hardcore TypeScript. I wish we used other patterns (render props, hooks) where types are more straighforward.